# IMPORT LIBRARY

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import accuracy_score, classification_report
from tensorflow.keras.models import Sequential
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.optimizers import Adam
from sklearn.preprocessing import LabelEncoder

# LOAD DATASET

In [ ]:
from google.colab import files

# upload a file
uploaded = files.upload()

# read the uploaded CSV file with proper encoding
for filename in uploaded.keys():
    df = pd.read_csv(filename, encoding="utf-8", keep_default_na=False)

# display the first 5 rows of the DataFrame
df.head()

Saving data_scraping_MLBB.csv to data_scraping_MLBB.csv


reviewId         userName  \
0  3a4b6aff-694e-4e58-a1b3-39446f8f095c  Pengguna Google   
1  79b4abc1-6896-4070-b6da-dca3dfa89154  Pengguna Google   
2  43e35fe2-27d3-4a1d-8144-cc7ee7fb6a1d  Pengguna Google   
3  beaf12ab-ef17-4c9e-bf5c-759365745474  Pengguna Google   
4  397d3bca-f19b-452c-a638-a5d8209f4f14  Pengguna Google   

                                           userImage  \
0  https://play-lh.googleusercontent.com/EGemoI2N...   
1  https://play-lh.googleusercontent.com/EGemoI2N...   
2  https://play-lh.googleusercontent.com/EGemoI2N...   
3  https://play-lh.googleusercontent.com/EGemoI2N...   
4  https://play-lh.googleusercontent.com/EGemoI2N...   

                                             content  score  thumbsUpCount  \
0  Magic chess mengalami bug, pabila sudah mendap...      1            117   
1  Perbaiki jaringan,perbaiki juga foto album mas...      2            306   
2  game nya si bagus , tapi kenapa pas saya main ...      1            286   
3  maaf saya ubah ranting 1 buat Sistem Zona Line...      1            564   
4  season ini kenapa? sinyal bagus data ngalirnya...      1            416   

  reviewCreatedVersion                   at  \
0         1.9.93.10904  2025-07-21 21:44:48   
1         1.9.93.10904  2025-07-17 17:27:36   
2         1.9.93.10904  2025-07-21 09:33:55   
3         1.9.93.10904  2025-07-17 17:04:13   
4         1.9.70.10634  2025-07-16 01:43:00   

                                        replyContent            repliedAt  \
0  Halo Kak,\nKami mohon maaf atas pengalaman kur...  2025-07-26 08:48:39   
1  Halo Kak,\nMaaf atas masalah jaringan yang Kak...  2025-07-20 07:46:48   
2  Halo Kak,\nMaaf atas masalah jaringan yang Kak...  2025-07-23 03:27:08   
3  Halo Kak,\nkami berkomitmen untuk menciptakan ...  2025-07-18 08:15:30   
4  Halo Kak,\nMaaf atas masalah jaringan yang Kak...  2025-07-19 03:13:42   

     appVersion                                         text_clean  \
0  1.9.93.10904  Magic chess mengalami bug pabila sudah mendapa...   
1  1.9.93.10904  Perbaiki jaringanperbaiki juga foto album masa...   
2  1.9.93.10904  game nya si bagus  tapi kenapa pas saya main s...   
3  1.9.93.10904  maaf saya ubah ranting  buat Sistem Zona Line ...   
4  1.9.70.10634  season ini kenapa sinyal bagus data ngalirnya ...   

                                text_casefoldingText  \
0  magic chess mengalami bug pabila sudah mendapa...   
1  perbaiki jaringanperbaiki juga foto album masa...   
2  game nya si bagus  tapi kenapa pas saya main s...   
3  maaf saya ubah ranting  buat sistem zona line ...   
4  season ini kenapa sinyal bagus data ngalirnya ...   

                                     text_slangwords  \
0  magic chess mengalami bug pabila sudah mendapa...   
1  perbaiki jaringanperbaiki juga foto album masa...   
2  game nya si bagus tapi kenapa pas saya main se...   
3  maaf saya ubah ranting buat sistem zona line y...   
4  season ini kenapa sinyal bagus data ngalirnya ...   

                                 text_tokenizingText  \
0  ['magic', 'chess', 'mengalami', 'bug', 'pabila...   
1  ['perbaiki', 'jaringanperbaiki', 'juga', 'foto...   
2  ['game', 'nya', 'si', 'bagus', 'tapi', 'kenapa...   
3  ['maaf', 'saya', 'ubah', 'ranting', 'buat', 's...   
4  ['season', 'ini', 'kenapa', 'sinyal', 'bagus',...   

                                       text_stopword  \
0  ['magic', 'chess', 'mengalami', 'bug', 'pabila...   
1  ['perbaiki', 'jaringanperbaiki', 'foto', 'albu...   
2  ['game', 'si', 'bagus', 'pas', 'main', 'lag', ...   
3  ['maaf', 'ubah', 'ranting', 'sistem', 'zona', ...   
4  ['season', 'sinyal', 'bagus', 'data', 'ngalirn...   

                                          text_akhir  polarity_score sentiment  
0  magic chess mengalami bug pabila hero gambarny...               0  positive  
1  perbaiki jaringanperbaiki foto album relog mul...             -16  negative  
2  game si bagus pas main lag sinyal hijau kuat p...              10  positive  
3  maaf ubah ranting sist

# FEATURE EXTRACTION

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# feature extraction using TF-IDF
X = df['text_akhir']  # preprocessed text
y = df['sentiment']   # sentiment labels

# initialize TF-IDF Vectorizer
tfidf = TfidfVectorizer(
    min_df=5,          # ignore terms that appear in fewer than 5 documents
    max_df=0.8,        # ignore terms that appear in more than 80% of documents
    sublinear_tf=True, # apply sublinear term frequency scaling
    use_idf=True       # enable inverse document frequency weighting
)

# transform the text data into TF-IDF features
X_tfidf = tfidf.fit_transform(X)

# convert the TF-IDF result to a DataFrame
features_df = pd.DataFrame(X_tfidf.toarray(), columns=tfidf.get_feature_names_out())

# display the extracted features
features_df

abadi  abang  acak  acara   ad  adain  adakan  adil  adilan  adill  ...  \
0       0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
1       0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
2       0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
3       0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
4       0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
...     ...    ...   ...    ...  ...    ...     ...   ...     ...    ...  ...   
9502    0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
9503    0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
9504    0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
9505    0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   
9506    0.0    0.0   0.0    0.0  0.0    0.0     0.0   0.0     0.0    0.0  ...   

       yh  yin  yng  youtube  yss   yt  yth  yuk  zaman  zilong  
0     0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
1     0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
2     0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
3     0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
4     0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
...   ...  ...  ...      ...  ...  ...  ...  ...    ...     ...  
9502  0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
9503  0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
9504  0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
9505  0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  
9506  0.0  0.0  0.0      0.0  0.0  0.0  0.0  0.0    0.0     0.0  

[9507 rows x 2773 columns]

# MACHINE LEARNING ALGORITHM

In [ ]:
from sklearn.model_selection import train_test_split

# split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X_tfidf,      # feature matrix
    y,            # target labels (sentiment)
    test_size=0.2,  # 20% of the data will be used for testing
    random_state=42 # ensures reproducibility
)

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y_train_enc = le.fit_transform(y_train)
y_test_enc = le.transform(y_test)

# ACCURACY TESTING SET

In [ ]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report

# create the MLP model
mlp_model = MLPClassifier(
    hidden_layer_sizes=(512, 256, 128),  # three hidden layers with 512, 256, and 128 neurons
    activation='relu',                  # ReLU activation function
    solver='adam',                      # adam optimizer
    alpha=1.0,                          # L2 regularization term
    learning_rate='adaptive',          # adaptive learning rate
    max_iter=1000,                      # maximum number of iterations
    early_stopping=True,               # stop early if validation score doesn't improve
    n_iter_no_change=10,               # stop after 10 epochs with no improvement
    random_state=42                    # seed for reproducibility
)

# train the model
mlp_model.fit(X_train, y_train_enc)

# predict on training and testing sets
y_pred_train_mlp = mlp_model.predict(X_train)
y_pred_test_mlp = mlp_model.predict(X_test)

# evaluate accuracy
accuracy_train_mlp = accuracy_score(y_train_enc, y_pred_train_mlp)
accuracy_test_mlp = accuracy_score(y_test_enc, y_pred_test_mlp)

# display evaluation results
print(f'MLP Training Accuracy: {accuracy_train_mlp:.4f}')
print(f'MLP Testing Accuracy : {accuracy_test_mlp:.4f}')
print('\nClassification Report (Testing - MLP):\n', classification_report(y_test_enc, y_pred_test_mlp))


MLP Training Accuracy: 0.9773
MLP Testing Accuracy : 0.8812

Classification Report (Testing - MLP):
               precision    recall  f1-score   support

           0       0.90      0.93      0.91      1268
           1       0.84      0.79      0.82       634

    accuracy                           0.88      1902
   macro avg       0.87      0.86      0.86      1902
weighted avg       0.88      0.88      0.88      1902



# INSIGHT

Pada notebook ini, saya melakukan analisis sentimen dari data scraping terkait penilaian aplikasi Mobile Legends pada PlayStore. Analisis ini menggunakan teknik Extraksi Fitur (TF IDF), Algoritma Machine Learning dan Akurasi Set Latihan. Hasil akurasi yang saya dapatkan pada notebook ini yaitu untuk Training Accuracy 97% dan Testing Accuracy 88%. Hasil ini menunjukkan bahwa model sudah cukup baik untuk mengenal pola sentimen tanpa overfitting.